src/test/scala/com/microsoft/hyperspace/index/rules/JoinIndexRuleV2Test.scala

```diff
@@ -0,0 +1,283 @@
+
+    t1ScanNode = LogicalRelation(t1Relation, Seq(t1c1, t1c2, t1c3, t1c4), None, false)
+    t2ScanNode = LogicalRelation(t2Relation, Seq(t2c1, t2c2, t2c3, t2c4), None, false)
+    t3ScanNode = LogicalRelation(t3Relation, Seq(t3c1, t3c2, t3c3, t3c4), None, false)
+
+    t1FilterNode = Filter(And(IsNotNull(t1c1), IsNotNull(t1c2)), t1ScanNode)
+    t2FilterNode = Filter(And(IsNotNull(t2c1), IsNotNull(t2c2)), t2ScanNode)
+    t3FilterNode = Filter(IsNotNull(t3c1), t3ScanNode)
+
+    t1ProjectNode = Project(Seq(t1c1, t1c2, t1c3), t1FilterNode)
+
+    t2ProjectNode = Project(Seq(t2c1, t2c2, t2c3), t2FilterNode)
+
+    t3ProjectNode = Project(Seq(t3c2, t3c3), t3FilterNode)
+
+    createIndexLogEntry("t1i1", Seq(t1c1), Seq(t1c2, t1c3), t1ProjectNode)
+    createIndexLogEntry("t1i2", Seq(t1c1, t1c2), Seq(t1c3), t1ProjectNode)
+    createIndexLogEntry("t2i1", Seq(t2c1), Seq(t2c2, t2c3), t2ProjectNode)
+    createIndexLogEntry("t2i2", Seq(t2c2), Seq(t2c1, t2c3), t2ProjectNode)
+    createIndexLogEntry("t2i3", Seq(t2c1, t2c2), Seq(t2c3), t2ProjectNode)
+    createIndexLogEntry("t3i1", Seq(t3c2), Seq(t3c3), t3ProjectNode)
+  }
+
+  before {
+    spark.conf.set(IndexConstants.INDEX_SYSTEM_PATH, systemPath.toUri.toString)
+    clearCache()
+  }
+
+  private def baseRelation(
+      location: FileIndex,
+      schema: StructType,
+      spark: SparkSession): HadoopFsRelation = {
+    HadoopFsRelation(location, new StructType(), schema, None, new ParquetFileFormat, Map.empty)(
+      spark)
+  }
+
+  override def afterAll(): Unit = {
+    FileUtils.delete(systemPath)
+    super.afterAll()
+  }
+
+  test("Join rule does not update plan for Broadcast Hash Join compatible joins.") {
+    withSQLConf("spark.hyperspace.rule.joinV2.enabled" -> "true") {
+      val joinCondition = EqualTo(t1c1, t3c1)
+      val originalPlan =
+        Join(t1ProjectNode, t3ProjectNode, JoinType("inner"), Some(joinCondition))
+      val updatedPlan = JoinIndexRuleV2(originalPlan)
+      assert(updatedPlan.equals(originalPlan))
+    }
+  }
+
+  test(
+    "Join rule updates sort merge join part of a plan if indexes exist and configs are set" +
+      " correctly.") {
+    withSQLConf("spark.hyperspace.rule.joinV2.enabled" -> "true") {
+      val joinCondition = EqualTo(t1c1, t2c1)
+      val originalPlan =
+        Join(t1ProjectNode, t2ProjectNode, JoinType("inner"), Some(joinCondition))
+      val updatedPlan = JoinIndexRuleV2(originalPlan)
+      assert(!updatedPlan.equals(originalPlan))
+
+      val indexPaths = getIndexDataFilesPaths("t1i1") ++ getIndexDataFilesPaths("t2i1")
+      verifyUpdatedIndex(originalPlan, updatedPlan, indexPaths)
+    }
+  }
+
+  test("Join rule updates sort merge join part of a plan with both smj and bhj.") {
+    // Test Setup:
+    //      SMJ(t1c1 = t2c1)
+    //     /   \
+    //   t1    BHJ(t2c2 = t3c2)
+    //        /   \
+    //       t2   t3
+    // When V2 is run, this should optimize indexes for t1 and t2 based on join condition in SMJ
```

Comment:
How about SMJ - SMJ ? What if there is a node with different output partitioning (like Union?) in the middle of left or right? Should we need to check if the bucketing info is preserved or no?
```
  SMJ
 /   \
t1    SMJ
      /  \
     t2  t3
```

Reply:
added test case in E2E test to explain behavior in such scenario. in summary, lower smj will be optimized.

Since I am replacing only the scan node with index (or Union in case of hybrid scan) I think the ensure requirements phase will make sure to introduce bucketing when required. Could you please give an example where this could fail? I couldn't come up with one.

Reply:
I meant - SMJ-SMJ case, if the lower SMJ isn't eligible? But it requires shuffle w/ different columns?

And for Union, I edited the above comment? 
Since v1 has isLinear check function, it's not a problem with v1, but I wonder how it handles in v2.

Reply:
> I meant - SMJ-SMJ case, if the lower SMJ isn't eligible? But it requires shuffle w/ different columns?

oh got it. then for lower SMJ, index won't be picked for any relation. If one of these relations is eligible for higher SMJ, it will be replaced by its index. I think this case is similar to a lower level BHJ scenario where build side is the side with supported index columns in higher level SMJ.
cc @imback82 

> And for Union, I edited the above comment.
> Since v1 has isLinear check function, it's not a problem with v1, but I wonder how it is handled in v2.
Union relations will remain untouched if the join columns used in higher level SMJ belong to Union node. If it belongs to only one of the tables t2 or t3, it will be replaced with the index. But if both t2 and t3 have some columns used in SMJ, we won't replace with index at all.

Bucketing info will not travel up the Union node because the other side of union is not bucketed.

Reply:
So in those cases, there should be a shuffle node even with the index, right?
I'm not sure if it's a good idea to apply indexes for them. 
And if Hybrid Scan is applied for these ineffective indexes, there should be some performance regression: 
- shuffle for BucketUnion (for hybrid scan, which is unnecessary)
- shuffle for join (higher SMJ)

Other than this, the change generally LGTM.

Reply:
> So in those cases, there should be a shuffle node even with the index, right?
> I'm not sure if it's a good idea to apply indexes for them.
> And if Hybrid Scan is applied for these ineffective indexes, there should be some performance regression:
>
> * shuffle for BucketUnion (for hybrid scan, which is unnecessary)

I thought you meant Union (not BucketUnion, which is different). In case of bucket union, which is added by hybrid scan, another extra shuffle node will not be added. So there should not be a perf degradation. (TBH, I don't think this case can occur in the first place)

> * shuffle for join (higher SMJ)

For SMJ I agree that that index will not be as effective, although reading index compared to the original data could have some benefits in terms of data read from disk. The same will be true for ALL kinds of Binary Nodes (which we have not yet worked with) and so I am not sure if using index in place of data (both will be shuffled anyway) will degrade performance definitely. So we have a couple of options. Let me know what you think:
- Add linearity check to join rule V2 (we don't use index for any kind of BinaryNode)
- Add checks for all kinds of Binary Nodes e.g. SMJ and others.

cc @imback82 , @pirz for suggestions.

Reply:
But `shuffle for join (for higherSMJ)` **AND** `shuffle for BucketUnion (for Hybrid Scan)` will cause degradation as the index is ineffective. Can we add linearity check at least for now? @imback82 @pirz

Reply:
> But `shuffle for join (for higherSMJ)` **AND** `shuffle for BucketUnion (for Hybrid Scan)` will cause degradation as the index is ineffective. Can we add linearity check at least for now? @imback82 @pirz

I don't think there will be another shuffle for higherSMJ. Only a shuffle within the BucketUnion as per my understanding of hybrid scan.
IF there is a BucketUnion, it means the Join Rule has added it. And it was added ONLY IF the children of bucket union node have same partitioning as higher level SMJ. Is that not correct? And if it is, there will NOT be any more shuffle node except for the one within the BucketUnion node

Reply:
I thought _not_ having the linearity check was helping few queries in TPC-DS where output partitioning from lower node needs to travel up? I wonder what would happen if we change this in terms of the perf gain. I guess we need the TPC-* test framework sooner than later to iterate quickly.

@apoorvedave1 Could you compile a list of concerns in #237 so we can better track (if it's missing any of the worthy discussion)?

Reply:
> And it was added ONLY IF the children of bucket union node have same partitioning as higher level SMJ. Is that not correct? And if it is, there will NOT be any more shuffle node except for the one within the BucketUnion node

In case the lower level SMJ doesn't retain the bucketing spec, there will be a shuffle for the higher level SMJ, and then shuffle for BucketUnion (for the higher level SMJ) will be a cause of perf regression. WDYT?

Reply:
I really got confused now. Could you create another diagram please? it looks like you have 2 SMJs and a Bucket union.

Reply:
Ok for example:
```
    SMJ
    /   \
 t1   SMJ (not eligible for index)
        /   \
shuffle   shuffle
     /       \
   t2       t3

will be transformed:

      SMJ1 (eligible for index, higher level SMJ)
      /   \
   t1     SMJ2 (not eligible for index, lower level SMJ, different bucketing columns)
        /           \
shuffle for SMJ2     shuffle - t3
      /  
 bucket union for SMJ1
  /    \
t2 idx   shuffle for SMJ1
          \
        t2 appended
```